### ISSUE TITLE
Tail Latency with Puma Queues in Single Threaded mode

### ISSUE THREAD

**Issue by danmayer** [puma/puma#2311]
**Describe the bug**

We have been investigating various latencies in our system for a while. We recently tried [Puma5 and the performance features](https://github.com/puma/puma/issues/2258#issuecomment-659118154), to see if it had any positive impact, but were able to find a more direct issue with Puma Queues. For reasons that I won't get into and that we are working to address we run Puma in single-threaded mode, and I suspect a large part of the issue is related to queues without having threads.

We see tail latency in the p95/p99 of 100s of ms related to using puma queues in single-threaded mode.

![Screen Shot 2020-07-15 at 8 27 51 PM](https://user-images.githubusercontent.com/24925/87619567-b8de4a00-c6d9-11ea-9989-d498dd70ec08.png)
> App with no changes other than turning puma queues back on, an immediate jump of 150-180ms on the p99 latency

**Puma Threads**

I was able to find a safe application to enable threading and the issue nearly goes away when one is running Puma in threaded mode with even 2 threads (`MAX_THREADS=2`). This is why much of the focus has been related to the impact on single-threaded application servers. 

**Puma config:**

This has been repeated on dozens of app with similar effects, although many have slightly different puma configs... This is a standard puma config example. A typical app would have `WEB_CONCURRENCY = 6` and `MAX_THREADS` not set so defaulting to 1. We can toggle puma queues with the ENV var `PUMA_QUEUE_REQUESTS`

```
workers Integer(ENV["WEB_CONCURRENCY"] || 2)

threads_count = Integer(ENV["MAX_THREADS"] || 1)
threads threads_count, threads_count

preload_app!

rackup DefaultRackup
port ENV["PORT"] || 3000
environment ENV["RACK_ENV"] || "development"

queue_requests(ENV["PUMA_QUEUE_REQUESTS"] == "true")

require "company/metrics/puma/plugin/puma_stats"
plugin :puma_stats
```

**To Reproduce**

I can try to work on a small example, but we don't see these kinds of impacts until under fairly sizable load on services with multiple servers. I can investigate if I could reproduce with a rack app, but at the moment the fix holds across a larger number of very different Rails apps.

**Expected behavior**

That queueing doesn't significantly impact latency...

**Data Collection Samples**

Each of the images includes the before and after of disabling puma queueing.

![Screen Shot 2020-07-15 at 8 45 48 PM](https://user-images.githubusercontent.com/24925/87620603-3c993600-c6dc-11ea-8fe3-f8d8b8c04c33.png)

**Impact on other related metrics**

The primary metric we cared about was client timed performance so the ALB or measure taken from API Client metrics. That is what I shared above, but...

We had other puma stats impacted

* puma.backlog basically goes to 0, which makes sense
* available pool capacity increases by say 30% 
* min pool capacity per server has a higher total (IE less server were running near max capacity)
* CPU utilization reduced
* Memory RSS reduced

** deployment machine:**
 - OS: ecs docker containers

Let me know other data collection, and information that would be helpful. I can help collect any metrics that might give insight into what is happening... I could add logs or possibly try experimental builds.

**Comment by ghiculescu** [comment]
I only just learned about the queueing config option. According to https://www.rubydoc.info/gems/puma/Puma%2FDSL:queue_requests it’s on by default. Does this mean it’s false in the majority of your screenshot, but true during the spike?

We run a similar lack of threads (though I think we are due for another attempt at thread safety soon). I can try turning queuing off and seeing if our results match yours.

**Comment by danmayer** [comment]
Yes @ghiculescu by default puma uses queues... Most of my graphs show the improvement when we disable it... The one graph that spikes up and then back down was when we tried to turning queuing back on. In all the graphs the only change was turning queueing off. If you are single-threaded, I would recommend giving that a shot as it had a significant impact for our services.

**Comment by nateberkopec** [comment]
It might be helpful to go back through the issue tracker/commit log and look at other issues re: queue_requests. The feature has a long history and a lot of perf changes/impact because it sits at a critical point in the response process.

For anyone reading this thread, please do be careful with this setting as running with `queue_requests false` means you are vulnerable to slow clients if you are not using a reverse proxy in front of Puma.

**Comment by danmayer** [comment]
yeah, @nateberkopec  it is a good point on the slow client, the majority of apps we have been making this change on are not on the public internet and are on private internal networks with controlled clients, so it isn't a concern we have to worry about, but many folks should ensure they are protected against slow client attacks. 

OK, I can try to look over the issues / commits related to queue_requests and see if there is anything interesting in there. One of the reasons we tried this was the comments around the toggle, which mention it might not be helpful in single threaded setups.

```
    # Queueing requests generally improves performance. In some
    # cases, such as a single threaded application, it may be
    # better to ensure requests get balanced across workers.
```

https://github.com/puma/puma/blob/5d02e45df59cb5ca2b318df471a9292688cfd021/lib/puma/dsl.rb#L678

**Comment by ghiculescu** [comment]
We've been running `queue_requests(false)` in our staging environment for the last day or so, and have not seen any negative results. It's mostly been fake scale though. We are going to test in some production environments today and report back.

**Comment by ghiculescu** [comment]
Deployed production with `queue_requests(false)`. Haven't seen any noticeable changes yet - definitely haven't replicated OP's issue. The only difference I noticed is we don't use `preload_app!`.

**Comment by nateberkopec** [comment]
Thanks for the repro attempt @ghiculescu. I don't think preloading would make a latency difference.

**Comment by danmayer** [comment]
Note for anyone else checking in we really only see these large improvements on backend internal services, so things that aren't exposed to the public internet and also don't have any frontend/asset pipeline/web pack... Basically, on pure restful JSON services. In case that helps anyone else that might want to try and reproduce.

Sorry you didn't see any benefit @ghiculescu and we believe the best long term solution is to resolve issues blocking us from enabling threading, so we are working on that more than trying to further debug this issue.

**Comment by ghiculescu** [comment]
I agree. What threading issues are you having? Would love to trade notes.

**Comment by dentarg** [comment]
I was searching for another issue and just happened to see this one... funny with the timing as https://github.com/rails/rails/issues/50450 just reached an conclusion – albeit that issue isn't discussing running Puma without any workers it does touch on the subject of tail latency, so I thought it would be good to connect these issues.

I'll also make a mental note to myself to re-read this issue when I have some time to do so.

**Comment by nateberkopec** [comment]
@dentarg funnily enough I think I reproduced the issue while making the benchmark for that thread. Note how [the benchmark](https://github.com/speedshop/threadbench) has to call `queue_requests false`. I did this because I was seeing atrocious performance when working with a threadpool of 1.

I suspect something is goofy regarding the Reactor thread accepting too many requests. It's possible that repro can only be done if the server is under significant utilization (80%+).

**Comment by danmayer** [comment]
We never solved this problem at my last gig we ran with `queue_requests false` or multiple threads... Spent some time on it but could never figure out a good way to debug it.

**Comment by dentarg** [comment]
Anyone here wanna try out https://github.com/puma/puma/pull/3506?

**Comment by dentarg** [comment]
https://github.com/puma/puma/pull/3678 was merged and released with [Puma 7.0.0.pre1](https://rubygems.org/gems/puma/versions/7.0.0.pre1), please try it out if you can!

**Comment by dentarg** [comment]
Please report any feedback in https://github.com/puma/puma/discussions/3697 so we have it all in one place.
